### PR TITLE
Scalability improvements to the redis cache module

### PR DESCRIPTION
**Pull request by naegelin**
Salt cache has a very inefficient means of processing salt mine data which can create heavy reads against the mine cache. For syndics which are in remote locations (different datacenters or high-latency hops) this means reads become slow and blocking in environments where mines are heavily used.  Significant performance improvements can be had by installing a local redis instance on or near the local syndic which is slave to the primary redis cache master (cache.redis.host).  This improvement allows salt to hit the local redis for all significant read activity while maintaining a synchronized cache across syndics.  In other words, using both the remote and local redis configuration options on a syndic ensures that cache writes are centralized and heavy read activity remains localized.

### CODE REVIEW

**Comment by naegelin** [comment]
I will leave this here for whoever wants to run with it as I don't have the bandwidth to meet some of these silly linting rules you guys have in place.

File: salt/cache/redis_cache.py [hunk 1]
```diff
@@ -37,73 +30,60 @@
...
         Most cloud hosted redis clusters will require this to be set to ``True``
-
 db: ``'0'``
     The database index.
-
     .. note::
-
         The database index must be specified as string not as integer value!
-
 password:
```
Comment (terminalmage) [inline salt/cache/redis_cache.py]: None of the lines above should have been deleted, many are there for Sphinx formatting. Please put them back.

File: salt/cache/redis_cache.py [hunk 2]
```diff
@@ -37,73 +30,60 @@
...
     Redis connection password.
-
```
Comment (terminalmage) [inline salt/cache/redis_cache.py]: This blank line needs to be put back.

File: salt/cache/redis_cache.py [hunk 3]
```diff
@@ -37,73 +30,60 @@
...
+    This option is used primarily on syndics which have a local redis slave (mirror) synced
+    to the primary master. Enabling this option will significantly increase cache performance
+    on syndics which have >1ms latency from the primary salt-master. Please note the database
+    number of the slave cannot be different from the database number on the primary redis.
+ro_host: ``localhost``
+    The hostname of the Redis read-only slave server.
+ro_port: ``6379``
+    The Redis server port of the read-only slave server.
+ro_password:
+    Redis connection password of the read-only slave.
```
Comment (terminalmage) [inline salt/cache/redis_cache.py]: Blank lines are needed after each of the above options.

File: salt/cache/redis_cache.py [hunk 4]
```diff
@@ -37,73 +30,60 @@
...
 Configuration Example:
-
 .. code-block::yaml
-
     cache.redis.host: localhost
```
Comment (terminalmage) [inline salt/cache/redis_cache.py]: These blank lines need to be added back.

File: salt/cache/redis_cache.py [hunk 5]
```diff
@@ -112,11 +92,27 @@
     cache.redis.bank_keys_prefix: #BANKEYS
     cache.redis.key_prefix: #KEY
     cache.redis.separator: '@'
-
```
Comment (terminalmage) [inline salt/cache/redis_cache.py]: This blank line needs to be added back.

File: salt/cache/redis_cache.py [hunk 6]
```diff
@@ -112,11 +92,27 @@
...
+    cache.redis.password: my pass
+    cache.redis.bank_prefix: #BANK
+    cache.redis.bank_keys_prefix: #BANKEYS
+    cache.redis.key_prefix: #KEY
+    cache.redis.separator: '@'
 Cluster Configuration Example:
-
 .. code-block::yaml
-
     cache.redis.cluster_mode: true
```
Comment (terminalmage) [inline salt/cache/redis_cache.py]: These blank lines need to be added back.

File: salt/cache/redis_cache.py [hunk 7]
```diff
@@ -112,11 +92,27 @@
...
     cache.redis.key_prefix: #KEY
     cache.redis.separator: '@'
-
+Syndic Configuration Example:
+    For syndics which are in remote locations (different datacenters or high-latency hops)
+    significant performance improvements can be had by installing a local redis slave
+    pointing to the primary redis cache master (cache.redis.host). Using both the remote
+    and local redis configuration options ensures that cache writes are centralized and
+    heavy read activity remains localized.
+.. code-block::yaml
```
Comment (terminalmage) [inline salt/cache/redis_cache.py]: A blank line is needed both before and after the `code-block` line.

File: salt/cache/redis_cache.py [hunk 8]
```diff
@@ -112,11 +92,27 @@
...
+    cache.redis.readonly_slave: True
+    cache.redis.ro_host: localhost
+    cache.redis.ro_port: 6379
+    cache.redis.ro_password: my pass
+    cache.redis.password: my pass
+    cache.redis.bank_prefix: #BANK
+    cache.redis.bank_keys_prefix: #BANKEYS
+    cache.redis.key_prefix: #KEY
+    cache.redis.separator: '@'
 Cluster Configuration Example:
```
Comment (terminalmage) [inline salt/cache/redis_cache.py]: Blank line needed before this line.

File: salt/cache/redis_cache.py [hunk 9]
```diff
@@ -381,14 +399,11 @@ def flush(bank, key=None):
...
     in order to re-create the tree, there are a couple of requests made. In total:
-
     - one for node in the hierarchy sub-tree, starting from the bank node
     - one pipelined request to get the keys under all banks in the sub-tree
     - one pipeline request to remove the corresponding keys
-
     This is not quite optimal, as if we need to flush a bank having
     a very long list of sub-banks, the number of requests to build the sub-tree may grow quite big.
-
     An improvement for this would be loading a custom Lua script in the Redis instance of the user
```
Comment (terminalmage) [inline salt/cache/redis_cache.py]: Add these blank lines back, please.

**Review by terminalmage — CHANGES_REQUESTED** [review]
You actually passed lint. We're still tuning the codeclimate check, so you can disregard it for now. You will however need to fix several issues with needlessly-removed blank lines which will affect Sphinx formatting.

**Comment by rallytime** [comment]
@naegelin Were you able to come back to the feedback here?

**Comment by rallytime** [comment]
I am going to close this since we haven't heard back here. We can always reopen when you're ready to come back to this. Or feel free to open a new PR. Thank you!

**Pull request closed without merging**